File: app/src/main/java/co/za/takeabyte/gradalot2020/deals/adapter/AdapterDeals.java
package co.za.takeabyte.gradalot2020.deals.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import co.za.takeabyte.gradalot2020.R;
import co.za.takeabyte.gradalot2020.promotions.listener.OnPromotionsListener;
import co.za.takeabyte.gradalot2020.promotions.uimodel.UIModelPromotionItem;
import co.za.takeabyte.gradalot2020.promotions.viewholder.ViewHolderPromotion;

public class AdapterDeals extends RecyclerView.Adapter<ViewHolderPromotion> {

    private List<UIModelPromotionItem> dataSet = new ArrayList<>();
    private OnPromotionsListener onPromotionsListener;

    public AdapterDeals(@NonNull OnPromotionsListener onPromotionsListener) {
        this.onPromotionsListener = onPromotionsListener;
    }

    @NonNull
    @Override
    public ViewHolderPromotion onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View itemView = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.deals_item_layout, parent, false);

        return new ViewHolderPromotion(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderPromotion holder, int position) {
        final UIModelPromotionItem model = dataSet.get(position);
        if (model != null) {
            holder.bindView(model);
            holder.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (onPromotionsListener != null) {
                        onPromotionsListener.onPromotionsItemSelected(model);
                    }
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return dataSet.size();
    }

    public void setDataSet(@NonNull List<UIModelPromotionItem> models) {
        dataSet = models;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/co/za/takeabyte/gradalot2020/deals/adapter/AdapterDeals.kt
package co.za.takeabyte.gradalot2020.deals.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import co.za.takeabyte.gradalot2020.R
import co.za.takeabyte.gradalot2020.promotions.uimodel.UIModelPromotionItem
import co.za.takeabyte.gradalot2020.promotions.viewholder.ViewHolderPromotion

class AdapterDeals(
    private val onPromotionItemSelected: (ImageView, UIModelPromotionItem) -> Unit
) : RecyclerView.Adapter<ViewHolderPromotion>() {

    var dataSet: List<UIModelPromotionItem> = emptyList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderPromotion {
        val itemView = LayoutInflater.from(parent.context).inflate(
            R.layout.deals_item_layout,
            parent,
            false
        )

        return ViewHolderPromotion(itemView)
    }

    override fun getItemCount() = dataSet.size

    override fun onBindViewHolder(holder: ViewHolderPromotion, position: Int) {
        val model = dataSet[position]
        holder.bindView(model, onPromotionItemSelected)
    }

}